### PR TITLE
Executer regression fix

### PR DESCRIPTION
**Pull request by jmachowinski**
Fixes https://github.com/ros2/rclcpp/issues/2508

### CODE REVIEW

File: rclcpp/test/rclcpp/executors/test_executors.cpp [hunk 1]
```diff
@@ -869,3 +856,75 @@ TEST(TestExecutors, testSpinWithNonDefaultContext)
...
+    start_time = std::chrono::system_clock::now();
+  }
+
+  void TearDown()
+  {
+    // there were ne events ready, a test iteration should come back in less than 100 ms
+    EXPECT_LT(
+      std::chrono::system_clock::now() - start_time,
+      max_duration / 2)
+      << "spin_some() took a long time to execute when it should have done "
```
Comment (fujitatomoya) [inline rclcpp/test/rclcpp/executors/test_executors.cpp:894]: nitpick: test includes `spin_all` as well.

```suggestion
      << "executor took a long time to execute when it should have done "
```

**Review by fujitatomoya — APPROVED** [review]
lgtm!

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20877/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15180/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21559/)

**Review by alsora — APPROVED** [review]

**Comment by jmachowinski** [comment]
I looked further into this, as all the tests lid up.
I pushed an updated version, can someone rerun the CI with this ?

@wjwwood @mjcarroll I noticed, that the spin_some_impl method did not act according to the documentation.
It was not recollecting ready executables on every call. Is this intentional, or a bug ?

**Comment by fujitatomoya** [comment]
@jmachowinski sorry i missed this, here is new CI.

CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20938/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15229/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21612/)

**Review by wjwwood — CHANGES_REQUESTED** [review]
I couldn't fix up this pr directly because of push permissions, but I made a new pr that builds on this one: https://github.com/ros2/rclcpp/pull/2517

I think we should take my pr, but I'm open to discussing even more changes.

**Comment by wjwwood** [comment]
Closing in favor of https://github.com/ros2/rclcpp/pull/2517, but we can reopen this one if needed.

File: rclcpp/src/rclcpp/executor.cpp [hunk 2]
```diff
@@ -374,16 +377,19 @@ Executor::spin_some_impl(std::chrono::nanoseconds max_duration, bool exhaustive)
...
       // If nothing is ready, reset the result to signal we are
       // ready to wait again
       wait_result_.reset();
-    }
 
-    if (!wait_result_.has_value() && !exhaustive) {
-      // In the case of spin some, then we can exit
-      // In the case of spin all, then we will allow ourselves to wait again.
-      break;
+      if (!work_available || !exhaustive) {
```
[the remark this answers is not in the document]
Comment (jmachowinski) [reply rclcpp/src/rclcpp/executor.cpp:386]: You would not end up in this code path as long as work is available, as get_next_ready_executable(any_exec) would return true.

**Pull request closed without merging**